clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -501,13 +501,16 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 }
 
                 if (exception instanceof UnknownMemberIdException ||
-                    exception instanceof IllegalGenerationException ||
-                    exception instanceof RebalanceInProgressException ||
-                    exception instanceof MemberIdRequiredException)
+                        exception instanceof IllegalGenerationException ||
+                        exception instanceof RebalanceInProgressException ||
+                        exception instanceof MemberIdRequiredException)
```

Comment:
Should we actually do the timer check before this? Since otherwise if the exception from the immediately returned responses is any of those four, we would still `continue` and skip the check below.

More concretely I think we can just move the remaining logic inside the `if` call:

```
if (!future.isRetriable()) {
    throw ..
} else {
    if (timer.isExpired()) {
        return false;
    } else if (exception instance of..) {
        continue;
    } else {
        timer.sleep(..)
    }
}
```

Reply:
agreed. A comment here, retriableException check should happen after the `instanceOf` checks, because I think we actually want to retry upon these (according to the logic).